### ISSUE TITLE
Reconnect can throw an unhandled exception

### ISSUE THREAD

**Issue by harrisa1** [nats-io/nats.net.v1#268]
During regular usage there was something that caused the nats connection to drop (i.e. a network blip).  After this the client tried to reconnect and when it did it hit an exception
```
Type=System.InvalidOperationException
Source=System.Collections
Message=Collection was modified; enumeration operation may not execute.
StackTrace:
   at System.Collections.Generic.Queue`1.Enumerator.MoveNext()
   at NATS.Client.Connection.clearPendingFlushCalls()
   at NATS.Client.Connection.doReconnect()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```
Looking at the above I am guessing that a ConcurrentQueue would probably serve better since the collection can be modified while enumerating.

**Comment by watfordgnf** [comment]
`pongs` appears to be covered by the `mu` lock. It would be good to know how `pongs` was altered outside of `mu`.

Found it:
https://github.com/nats-io/nats.net/blob/012ce23603b384dd422735e33b10f12c8e3309e7/NATS.Client/Conn.cs#L3557-L3562

`removeFlushEntry` should be protected by `mu` but is not.

**Comment by RobinsonWM** [comment]
We encountered a crash today that I believe has the same root cause but a different exception.  I wanted to paste it here in case anyone else is searching for this issue.
It is in the same method (`clearPendingFlushCalls()`) as the OP's issue, and it is also caused by access to the `pongs` queue while another thread is mutating it, just from a different line.

```no-highlight
Type=System.IndexOutOfRangeException
Source=System.Private.CoreLib
Message=Index was outside the bounds of the array.
StackTrace:
   at System.Array.Clear(Array array, Int32 index, Int32 length)
   at System.Collections.Generic.Queue`1.Clear()
   at NATS.Client.Connection.clearPendingFlushCalls()
   at NATS.Client.Connection.doReconnect()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```

**Comment by ColinSullivan1** [comment]
@RobinsonWM , thank you for updating the issue.  This will be addressed soon...

**Comment by ColinSullivan1** [comment]
I think we need locking inside the single use channel (which then might makes sense to go back to the old channel).  The Flush API releases the conn lock, but flush is allowed to be used across multiple threads - so if there are multiple calls to flush, there would be access errors here.  @watfordgnf , wdyt?

**Comment by ColinSullivan1** [comment]
Resolved by #312
